src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeEventProvider.cs

```diff
@@ -69,6 +71,30 @@ uint IEventProvider.EventUnregister(long registrationHandle)
                 }
                 EventPipeInternal.WriteEventData(eventHandle, userData, (uint)userDataCount, activityId, relatedActivityId);
             }
+            else
```

Comment:
This solution doesn't give me a great feeling. A few things that trouble me:
- All other metadata registration happens in EventSource so it seems odd that one special case occurs here
- If we aren't already using event id 0 its easy to imagine that code somewhere has taken a dependency that ID 0 can be used as a sentinel value as an invalid id
- We aren't respecting the EventLevel. Whatever the level is the first time this API is called will be cached and all future invocations ignore the level that was passed in to reuse the original level. I am guessing this isn't solvable without using different event ids for different levels which further complicates using a hard-coded zero id.
- We have support in TraceLoggingEventTypes to dynamically create new metadata/descriptors so it feels awkward to create a 2nd dynamic path. I think it is legitimate for EventPipe to create TraceLoggingEventTypes even when we aren't operating in 'SelfDescribing' mode. This will give you event ids taken from an auto incrementing counter and no need to expose new public APIs in the metadata generator.

Reply:
>     * All other metadata registration happens in EventSource so it seems odd that one special case occurs here

I originally was trying to put this code in EventSource, but the issue I kept running in to is the creation logic is all built under the assumption that it can bail at any point and just give up. That makes it pretty difficult to make sure the event is ever created. I suppose I could add this logic to WriteEventString if this is the route that we end up taking.
 
>     * If we aren't already using event id 0 its easy to imagine that code somewhere has taken a dependency that ID 0 can be used as a sentinel value as an invalid id

Event ID 0 is already reserved for WriteEventString see the code here https://github.com/dotnet/runtime/blob/a1aca9d0039b7dfacfce69f291f2b801c3a439cf/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L3159-L3163

>     * We aren't respecting the EventLevel. Whatever the level is the first time this API is called will be cached and all future invocations ignore the level that was passed in to reuse the original level. I am guessing this isn't solvable without using different event ids for different levels which further complicates using a hard-coded zero id.

WriteEventString is poorly named. It's only used for writing error messages from EventSources, it is only called from ReportOutOfBandMessage and always has id=0 level=logalways keywords=-1. It should be named WriteErrorString or something like that.

>     * We have support in TraceLoggingEventTypes to dynamically create new metadata/descriptors so it feels awkward to create a 2nd dynamic path. I think it is legitimate for EventPipe to create TraceLoggingEventTypes even when we aren't operating in 'SelfDescribing' mode. This will give you event ids taken from an auto incrementing counter and no need to expose new public APIs in the metadata generator.

This is a good point. My goal was to try and mimic the existing WriteEventString behavior for etw as closely as possible. We could change it to just use trace logging style events and skip all of this business. I am not sure if that qualifies as a breaking change, but it seems like since this path is only for error messages it should be fine.

How would you feel about ripping out basically this whole method and just having a trace logging style event for any error messages?

Reply:
Thanks! Knowing that level/keyword are fixed and zero was set aside for this purpose makes the path seem somewhat better. What if we do this?
1. Move the code that calls MetadataGenerator and caches the handle from the provider up to WriteEventString
2. Remove the level/leyword parameters on WriteEventString to make it clear they are fixed values
3. Remove the id != 0 check here so the event won't be ignored.

I'd be OK with a TraceLogging approach too, but since we couldn't use it for manifested ETW it would be a little odd if ETW listeners get a by-design event id 0 and EventPipe listeners get a different event id.

Reply:
Ok, I just updated with your feedback if you could take a look again